keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,137 @@
+
+#### RestartPolicy of Pods as OnFailure
+
+When the Pods of the Job have `.spec.template.spec.restartPolicy = "OnFailure"`, the failed Pod stays on the node and the failed container within the Pod is restarted. The Pod might run indefinitely or get marked as `Failed` after the `.spec.activeDeadlineSeconds` is completed by Job if specified. In this case also, Kueue won't reclaim the resources of the Job until the Job gets completed as either `Failed` or `Succeeded`.
+
+Hence, as seen from the above discussed Pod failure cases, we conclude that the Pods of a Job which fail during its execution, its resources are not immediately reclaimed by Kueue. Only when the Job gets marks as `Failed`, the resources of failed Pods will be reclaimed by Kueue.
+
+## API
+
+A new field `ReclaimedPodSets` is added to the `.status` of Workload API.
+
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+    // conditions hold the latest available observations of the Workload
+    // current state.
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    Conditions []WorkloadCondition `json:"conditions,omitempty"`
+
+    // list of count of Pods of a PodSet with resources reclaimed
+    // +optional
+    ReclaimedPodSets []ReclaimedPodSetPods `json:"reclaimedPodSets"`
+}
+
+// ReclaimedPodSetPod defines the PodSet name and count of successfully completed Pods 
+// belonging to the PodSet whose resources can be reclaimed.
+type ReclaimedPodSet struct{
+    Name string
+    Count int
+}
+```
+
+## Implementation
+
+Kueue's job reconciler will compare the Job's `.status.Succeeded` field and the Workload's `.status.reclaimedPodSets[i].count` field value. If the former value is greater than the later then, Kueue's Job reconciler will update the Workload object's `.status`. Workload reconciler will catch the update event of Kueue Job reconciler and release the resources of the newly succeeded Pods to the ClusterQueue depending upon the case the Job satisfies discussed in the section of [successful Pod completion](#pod-successful-completion-cases).
+
+## Testing Plan
+
+Dynamically reclaiming resources enhancement has unit and integration tests. These tests
+are run regularly as a part of Kueue's prow CI/CD pipeline.
+
+### Unit Tests
+
+All the Kueue's core components must be covered by unit tests.  Here is a list of unit tests required for the modules of the feature:
```

Comment:
We might also need unit tests in the job controller. But we can decide that later, given that all tests are currently at integration level.